File: Source/Aladdin.ASN1/ASN1/Choice.cs
using System;
using System.IO;

namespace Aladdin.ASN1
{
	///////////////////////////////////////////////////////////////////////////
	// Выбор из множества объектов
	///////////////////////////////////////////////////////////////////////////
	public class Choice : IObjectFactory 
	{
		// конструктор при раскодировании
		public Choice(ObjectInfo[] info) { this.info = info; } private ObjectInfo[] info; 

		// проверить допустимость типа
		public bool IsValidTag(Tag tag)
		{
			// для всех возможных альтернатив
			foreach (ObjectInfo item in info)
			{
				// проверить совпадение типа
				if (item.IsValidTag(tag)) return true; 
			}
			return false; 
		}
		// проверить корректность объекта
		public void Validate(IEncodable encodable, bool encode)
		{
			// для всех возможных альтернатив
			foreach (ObjectInfo item in info)
			{
				// проверить совпадение типа
				if (!item.IsValidTag(encodable.Tag)) continue; 

				// проверить корректность объекта
				item.Validate(encodable, encode); return; 
			}
		    // выбросить исключение 
		    if (encode) throw new ArgumentException(); else throw new InvalidDataException(); 
		}
		// получить информацию кодирования
		public ObjectInfo GetEncodableInfo(IEncodable encodable) 
		{
			// для всех возможных альтернатив
			foreach (ObjectInfo item in info)
			{
				// проверить совпадение типа
				if (item.IsValidTag(encodable.Tag)) return item;
			}
			// ошибка - некорректный объект
			throw new InvalidDataException(); 
		}
        // раскодировать объект
        public IEncodable Decode(IEncodable encodable) 
        { 
			// для всех возможных альтернатив
			foreach (ObjectInfo item in info)
			{
				// проверить совпадение типа
				if (!item.IsValidTag(encodable.Tag)) continue; 

				// раскодировать объект
				return item.Decode(encodable, false);
			}
			// ошибка - некорректный объект
			throw new InvalidDataException(); 
        }
	}
}

File: Source/Aladdin.ASN1.Java/src/aladdin/asn1/Choice.java
package aladdin.asn1;
import java.io.*; 

///////////////////////////////////////////////////////////////////////////
// Выбор из множества объектов
///////////////////////////////////////////////////////////////////////////
public class Choice implements IObjectFactory<IEncodable>
{
    private final ObjectInfo[] info;
    
    // конструктор при раскодировании
    public Choice(ObjectInfo[] info) { this.info = info; } 

    // проверить допустимость типа
    @Override public final boolean isValidTag(Tag tag)
    {
        // для всех возможных альтернатив
		for (ObjectInfo item : info)
		{
            // проверить совпадение типа
            if (item.isValidTag(tag)) return true;
        }
		return false;
    }
    // раскодировать объект
    @Override public final IEncodable decode(IEncodable encodable) throws IOException
    {
        // для всех возможных альтернатив
		for (ObjectInfo item : info)
		{
            // проверить совпадение типа
            if (!item.isValidTag(encodable.tag())) continue;

            // раскодировать объект
            return item.decode(encodable, false);
		}
		// ошибка - некорректный объект
		throw new IOException();
    }
    // проверить корректность объекта
    @Override public final void validate(IEncodable encodable, boolean encode) throws IOException
    {
    	// для всех возможных альтернатив
    	for (ObjectInfo item : info)
        {
            // проверить совпадение типа
            if (!item.isValidTag(encodable.tag())) continue;

            // проверить корректность объекта
            item.validate(encodable, encode); return; 
        }
		// ошибка - некорректный объект
		if (encode) throw new IllegalArgumentException(); else throw new IOException();
    }
}